src/compiler/factory/utilities.ts

```diff
@@ -815,18 +815,118 @@ namespace ts {
+        return kind === SyntaxKind.AsteriskToken
+            || kind === SyntaxKind.SlashToken
+            || kind === SyntaxKind.PercentToken;
+    }
+
+    function isMultiplicativeOperatorOrHigher(kind: SyntaxKind): kind is MultiplicativeOperatorOrHigher {
+        return isExponentiationOperator(kind)
+            || isMultiplicativeOperator(kind);
+    }
+
+    function isAdditiveOperator(kind: SyntaxKind): kind is AdditiveOperator {
+        return kind === SyntaxKind.PlusToken
+            || kind === SyntaxKind.MinusToken;
+    }
+
+    function isAdditiveOperatorOrHigher(kind: SyntaxKind): kind is AdditiveOperatorOrHigher {
+        return isAdditiveOperator(kind)
+            || isMultiplicativeOperatorOrHigher(kind);
+    }
+
+    function isShiftOperator(kind: SyntaxKind): kind is ShiftOperator {
+        return kind === SyntaxKind.LessThanLessThanToken
+            || kind === SyntaxKind.GreaterThanGreaterThanToken
+            || kind === SyntaxKind.GreaterThanGreaterThanGreaterThanToken;
+    }
+
+    function isShiftOperatorOrHigher(kind: SyntaxKind): kind is ShiftOperatorOrHigher {
+        return isShiftOperator(kind)
+            || isAdditiveOperatorOrHigher(kind);
+    }
+
+    function isRelationalOperator(kind: SyntaxKind): kind is RelationalOperator {
+        return kind === SyntaxKind.LessThanToken
+            || kind === SyntaxKind.LessThanEqualsToken
+            || kind === SyntaxKind.GreaterThanToken
+            || kind === SyntaxKind.GreaterThanEqualsToken
+            || kind === SyntaxKind.InstanceOfKeyword
+            || kind === SyntaxKind.InKeyword;
+    }
+
+    function isRelationalOperatorOrHigher(kind: SyntaxKind): kind is RelationalOperatorOrHigher {
+        return isRelationalOperator(kind)
+            || isShiftOperatorOrHigher(kind);
+    }
+
+    function isEqualityOperator(kind: SyntaxKind): kind is EqualityOperator {
+        return kind === SyntaxKind.EqualsEqualsToken
+            || kind === SyntaxKind.EqualsEqualsEqualsToken
+            || kind === SyntaxKind.ExclamationEqualsToken
+            || kind === SyntaxKind.ExclamationEqualsEqualsToken;
+    }
+
+    function isEqualityOperatorOrHigher(kind: SyntaxKind): kind is EqualityOperatorOrHigher {
+        return isEqualityOperator(kind)
+            || isRelationalOperatorOrHigher(kind);
+    }
+
+    function isBitwiseOperator(kind: SyntaxKind): kind is BitwiseOperator {
+        return kind === SyntaxKind.AmpersandToken
+            || kind === SyntaxKind.BarToken
+            || kind === SyntaxKind.CaretToken;
+    }
+
+    function isBitwiseOperatorOrHigher(kind: SyntaxKind): kind is BitwiseOperatorOrHigher {
+        return isBitwiseOperator(kind)
+            || isEqualityOperatorOrHigher(kind);
+    }
+
+    // NOTE: The version in utilities includes ExclamationToken, which is not a binary operator.
+    function isLogicalOperator(kind: SyntaxKind): kind is LogicalOperator {

```

Comment:
... does `??` count nowadays (making our type inaccurate, too)? Or is it's precedence not similar to those? Looking at the [spec](https://tc39.es/ecma262/#sec-binary-logical-operators), I _think_ it should be included in the `LogicalOperator` group. Probably.

Reply:
`??` isn't a "logical" operator, per-se. It has a lower precedence than `&&` and `||`: https://tc39.es/ecma262/#prod-CoalesceExpression